doc/cpp17-features.md

```diff
@@ -0,0 +1,1113 @@
+[p0272r1](http://wg21.link/p0272r1)
+
+### 27. Multithreading and Concurrency
+
+#### 27.1. Supplementary Mutexes and Locks
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+HotSpot has its own threading mechanisms and mutexes, and does not use the
+Standard Library in this area.
+
+[n4470](http://wg21.link/n4470)
+[p0156r0](http://wg21.link/p0156r0)
+[p0156r2](http://wg21.link/p0156r2)
+[p0739r0](http://wg21.link/p0739r0)
+
+[n2406](http://wg21.link/n2406)
+[n4508](http://wg21.link/n4508)
+
+#### 27.2. `is_always_lock_free` for Atomics
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+HotSpot has its own atomics, and does not use the Standard Library in this area.
+
+[n4509](http://wg21.link/n4509)
+[p0152r1](http://wg21.link/p0152r1)
+
+#### 27.3. Cache Line Sizes
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+HotSpot has its own mechanisms for this. It's not clear that switching to
+these standard mechanisms is worthwhile. Quoting from the Josuttis book:
+
+"... if you know better, use specific values, but using these values is better
+than any assumed fixed size for code supporting multiple platforms."
+
+[n4523](http://wg21.link/n4523)
+[p0154r1](http://wg21.link/p0154r1)
+
+### 28. Other Small Library Features and Modifications
+
+#### 28.1. `std::uncaught_exceptions()`
+
+Not useful, since HotSpot is built with exceptions disabled.
+
+[n3614](http://wg21.link/n3614)
+[n4259](http://wg21.link/n4259)
+
+#### 28.2. Shared Pointer Improvements
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+[n3640](http://wg21.link/n3640)
+[p0414r2](http://wg21.link/p0414r2)
+[n4537](http://wg21.link/n4537)
+[p0163r0](http://wg21.link/p0163r0)
+[p0033r0](http://wg21.link/p0033r0)
+[p0033r1](http://wg21.link/p0033r1)
+
+#### 28.3. Numeric Extensions
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+[n3845](http://wg21.link/n3845)
+[p0295r0](http://wg21.link/p0295r0)
+[p0030r0](http://wg21.link/p0030r0)
+[p0030r1](http://wg21.link/p0030r1)
+[n1422](http://wg21.link/n1422)
+[p0226r1](http://wg21.link/p0226r1)
+
+#### 28.4. `chrono` Extensions
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+[p0092r0](http://wg21.link/p0092r0)
+[p0092r1](http://wg21.link/p0092r1)
+
+### 29. Polymorphic Memory Resources
+
+`FORBIDDEN_STANDARD_LIBRARY`
+
+If we ever start using Standard Library containers, we might base the
+allocators we provide on this set of features and use the corresponding
+containers.
+
+[n3525](http://wg21.link/n3525)
+[n3916](http://wg21.link/n3916)
+[p0220r1](http://wg21.link/p0220r1)
+
+### 34. Common C++ Settings
+
+#### 34.1. Value of `__cplusplus`
+
+We build HotSpot (and the rest of the OpenJDK) with a specifically selected
+version of the Standard. Hence, the value of `__cplusplus` should be known and
+unchanging until we change the project's build configuration again. So testing
+the value shouldn't ever be necessary.
+
+### 35. Deprecated and Removed Features
+
+#### 35.1. Deprecated and Removed Core Language Features
+
+##### 35.1.2. Keyword `register`
+
+We don't use the `register` keyword in the OpenJDK anymore?
```

Comment:
It's essential in some uses of GCC extended asm. See `atomic_linux_aarch64.hpp` for a good example.

Reply:
`register` is not used as a storage class specifier in the JDK. That use for
gcc extended asm is a gcc extension, with a different meaning and syntax. (It
has some similarities in appearance, but is really pretty different.) C++17
still retains `register` as a reserved word, so one still can't start using it
for other purposes.